src/redisearch_rs/rqe_iterators/src/optional.rs

```diff
@@ -0,0 +1,193 @@
+        debug_assert!(doc_id > self.result.doc_id);
+
+        if doc_id > self.max_doc_id || self.at_eof() {
+            self.result.doc_id = self.max_doc_id;
+            return Ok(None);
+        }
+
+        if let Some(child) = &mut self.child
+            && doc_id > child.last_doc_id()
+            && let Some(SkipToOutcome::Found(real)) = child.skip_to(doc_id)?
+            && real.doc_id == doc_id
+        {
+            // Only a "real" hit where a non-aborted child was able to skip
+            // to the desired doc_id (in contrast to being EOF before)
+            // reaches this branch
+            real.weight = self.weight;
+            self.result.doc_id = real.doc_id;
+            return Ok(Some(SkipToOutcome::Found(real)));
+        }
+
+        self.result.doc_id = doc_id;
+        Ok(Some(SkipToOutcome::Found(&mut self.result)))
+    }
+
+    fn revalidate(&mut self) -> Result<RQEValidateStatus<'_, 'index>, RQEIteratorError> {
+        let Some(ref mut child) = self.child else {
+            return Ok(RQEValidateStatus::Ok);
+        };
+        let last_child_doc_id = child.last_doc_id();
+
+        // Revalidate the child iterator (C-Code: step 1)
+        match child.revalidate()? {
+            // Abort: Handle child validation results (but continue processing)
+            // C-Code: step 2
+            RQEValidateStatus::Aborted => {
+                self.child = None; // Drop it so we become fully virtual until max is reached
+                Ok(if last_child_doc_id != self.result.doc_id {
+                    // virtual
+                    self.result.doc_id += 1;
+                    RQEValidateStatus::Moved {
+                        current: Some(&mut self.result),
+                    }
+                } else {
+                    // real
+                    RQEValidateStatus::Ok
+                })
+            }
+            // If the current result is virtual,
+            // or if the child was not moved, we can return VALIDATE_OK
+            //
+            // C-Code: step 3
+            RQEValidateStatus::Ok => Ok(RQEValidateStatus::Ok),
+            RQEValidateStatus::Moved { .. } => {
+                if last_child_doc_id != self.result.doc_id {
+                    // current result is virtual => VALIDATE_OK
+                    return Ok(RQEValidateStatus::Ok);
+                }
+
+                // Current result is real and child was moved - we need to re-read
+                //
+                // C-Code: step 4
+                Ok(RQEValidateStatus::Moved {
+                    current: self.read()?,
+                })
+            }
+        }
+    }
+
+    fn rewind(&mut self) {
+        self.result.doc_id = 0;
+        if let Some(child) = &mut self.child {
+            child.rewind();
+        }
+    }
+
+    fn num_estimated(&self) -> usize {
+        self.max_doc_id as usize
+    }
+
+    fn last_doc_id(&self) -> t_docId {
```

Comment:
Looks like it's not tested: https://app.codecov.io/gh/RediSearch/RediSearch/pull/7046/blob/src/redisearch_rs/rqe_iterators/src/optional.rs?dropdown=coverage#L186